Mathlib/FieldTheory/PurelyInseparable/Exponent.lean

```diff
@@ -0,0 +1,195 @@
+
+open Classical in
+variable {K} in
+theorem elemExponent_eq_zero_of_mem_range {a : L} (h : a ∈ (algebraMap K L).range) :
+    elemExponent K a = 0 := by
+  apply (Nat.find_eq_zero _).mpr
+  rw [pow_zero, pow_one]
+  obtain ⟨y, hy⟩ := h
+  exact ⟨y, hy ▸ minpoly.eq_X_sub_C L y⟩
+
+theorem elemExponent_eq_zero_of_charZero (a : L) [CharZero K] :
+    elemExponent K a = 0 :=
+  elemExponent_eq_zero_of_mem_range <| surjective_algebraMap_of_isSeparable K L a
+
+open Classical in
+/-- The element `y` of the base field `K` such that
+`a ^ ringExpChar K ^ elemExponent K a = algebraMap K L y`.
+See `IsPurelyInseparable.algebraMap_elemReduct_eq`. -/
+noncomputable def elemReduct (a : L) : K :=
+  Classical.choose <| Nat.find_spec <| minpoly_eq_X_pow_sub_C K (ringExpChar K) a
+
+open Classical in
+theorem minpoly_eq (a : L) :
+    minpoly K a = X ^ ringExpChar K ^ elemExponent K a - C (elemReduct K a) :=
+  Classical.choose_spec <| Nat.find_spec <| minpoly_eq_X_pow_sub_C K (ringExpChar K) a
+
+open Classical in
+/-- Version of `minpoly_eq` using `ExpChar`. -/
+theorem minpoly_eq' (p : ℕ) [ExpChar K p] (a : L) :
+    minpoly K a = X ^ p ^ elemExponent K a - C (elemReduct K a) :=
+  ringExpChar.eq K p ▸ minpoly_eq K a
+
+/-- The degree of the minimal polynomial of an element `a ∈ L` equals
+`ringExpChar K ^ elemExponent K a`. -/
+theorem minpoly_natDegree_eq (a : L) :
+    (minpoly K a).natDegree = ringExpChar K ^ elemExponent K a := by
+  rw [minpoly_eq K a, natDegree_sub_C, natDegree_pow, natDegree_X, mul_one]
+
+theorem algebraMap_elemReduct_eq (a : L) :
+    algebraMap K L (elemReduct K a) = a ^ ringExpChar K ^ elemExponent K a := by
+  have := minpoly_eq K a ▸ minpoly.aeval K a
+  rwa [map_sub, aeval_C, map_pow, aeval_X, sub_eq_zero, eq_comm] at this
+
+theorem elemExponent_def (a : L) :
+    a ^ ringExpChar K ^ elemExponent K a ∈ (algebraMap K L).range :=
+  RingHom.mem_range.mpr <| ⟨_, algebraMap_elemReduct_eq K a⟩
+
+variable {K} in
+theorem elemExponent_le_of_pow_mem {a : L} {n : ℕ}
+    (h : a ^ ringExpChar K ^ n ∈ (algebraMap K L).range) : elemExponent K a ≤ n := by
+  let ⟨p, _⟩ := ExpChar.exists K
+  rcases ‹ExpChar K p› with _ | ⟨hp⟩
+  · exact elemExponent_eq_zero_of_charZero K a ▸ Nat.zero_le _
+  · obtain ⟨y, hy⟩ := RingHom.mem_range.mp <| h
+    let f := X ^ ringExpChar K ^ n - C y
+    have hf₁ : f.aeval a = 0 := by rwa [map_sub, aeval_C, aeval_X_pow, sub_eq_zero, eq_comm]
+    have hf₂ : f.Monic := monic_X_pow_sub_C y <| Nat.pos_iff_ne_zero.mp <| expChar_pow_pos K _ _
+    have hf₃ : f.natDegree = ringExpChar K ^ n := by
+      rw [natDegree_sub_C, natDegree_pow, natDegree_X, mul_one]
+    exact (Nat.pow_le_pow_iff_right <| Nat.Prime.one_lt hp).mp <|
+      ringExpChar.eq K p ▸ hf₃ ▸ minpoly_natDegree_eq K a ▸
+      natDegree_le_natDegree (minpoly.min K a hf₂ hf₁)
+
+theorem elemExponent_min (a : L) {n : ℕ} (h : n < elemExponent K a) :
+    a ^ ringExpChar K ^ n ∉ (algebraMap K L).range :=
```

Comment:
Please also provide variants for these lemmas. After that, this seems ready for merging.

bors d+